internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,229 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ControllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+// ServiceAccountReconciler Reconciles for ServiceAccounts
+type ServiceAccountReconciler struct {
+	cfg ReconcilerConfig
+	client.Client
+}
+
+// SetupReconcilerWithManager initializes a reconciler
+// and registers that reconciler with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldSA := e.ObjectOld.(*corev1.ServiceAccount)
+				newSA := e.ObjectNew.(*corev1.ServiceAccount)
+				return hasControllerLabel(oldSA) != hasControllerLabel(newSA)
+			},
+			GenericFunc: func(event.GenericEvent) bool {
+				return false
+			},
+		}).
+		Complete(&ServiceAccountReconciler{
+			Client:         kargoMgr.GetClient(),
+		})
+}
+
+// Reconcile handles the reconciliation logic for ServiceAccounts
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceAccount", req.NamespacedName.Name,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	logger.Debug("Reconciling ServiceAccounts")
+
+	// Fetch the ServiceAccount
+	sa := &corev1.ServiceAccount{}
+	if err := r.Get(ctx, req.NamespacedName, sa); err != nil {
+		if kubeerr.IsNotFound(err) {
+			logger.Debug("ServiceAccount not found, deleting role bindings for all projects")
+			// If not found it means the ServiceAccount with the 
+			// controller label was deleted, hence remove Controller
+			// Permissions for the ServiceAccount.
```

Comment:
This comment isn't true because you didn't include any label-based criteria in your `Get()`. Also, I _wouldn't_ add label-based criteria to the `Get()`. At this stage, we don't know if the SA exists or doesn't. If it does, we don't know whether it has the label or not.

We should be getting the ServiceAccount exactly as you do now, drop this comment, and then deal with the following possibilities:

1. Not found --> ensure binding doesn't exist. (This happens if final deletion occurred in-between a create or update event being enqueued and the time the reconciler actually begins processing the reconcilation request.)

1. Found and lacks label OR deletion timestamp is not nil --> ensure the binding doesn't exist; ensure kargo finalizer is cleared.

1. Else (found, has label, has nil deletion timestamp) --> ensure finalizer exists; ensure binding exists.